### ISSUE TITLE
Design core API specification

### ISSUE THREAD

**Issue by ashleyvega** [vegaprotocol/specs#130]
In GitLab by @tamlyn10 on Sep 10, 2019, 15:39

# Description

Core API specs should ensure that all 'raw protocol data' is exposed in APIs. 
This includes data used in the protocol such as Order Id, parties, markets etc.
This excludes "calculated" quantities such as "closed PnL".

**Comment by ashleyvega** [comment]
In GitLab by @tamlyn10 on Sep 10, 2019, 15:39

assigned to @ashleyvega

**Comment by ashleyvega** [comment]
In GitLab by @tamlyn10 on Sep 10, 2019, 15:39

assigned to @barnabee

**Comment by ashleyvega** [comment]
In GitLab by @tamlyn10 on Sep 10, 2019, 15:53

changed the description

**Comment by ashleyvega** [comment]
In GitLab by @tamlyn10 on Sep 10, 2019, 15:56

@ashleyvega @barnabee I have raised an MR with a rough drafted version of Barney's thoughts how to structure this spec.

https://gitlab.com/vega-protocol/product/merge_requests/15

@ashleyvega would you be willing to iterate this - asking for any help you need from either of us? If you're too busy, no worries I'll take it on!

Not sure if @campbellssource or @eddhannay want to be involved?

**Comment by ashleyvega** [comment]
In GitLab by @ashleyvega on Sep 25, 2019, 14:58

@tamlyn10 @cdm I'm looking at this now...

**Comment by ashleyvega** [comment]
In GitLab by @ashleyvega on Sep 25, 2019, 17:14

mentioned in commit 1d3495f6dbfc2c198b6488cf0dfa5c82bde33d53

**Comment by ashleyvega** [comment]
In GitLab by @ashleyvega on Sep 30, 2019, 15:55

assigned to @cdm and unassigned @ashleyvega

**Comment by ashleyvega** [comment]
In GitLab by @cdm on Oct 3, 2019, 13:07

Small update from me, I've spoken with Tamlyn a couple of times on this and I'm re-writing the core API spec document should be pushed to a branch later today.

Main focus for this spec is to clarify what we think of as the 'core' APIs and provide an overview of the expectation. 

I'd like to propose that the APIs for a feature of the protocol are described with the feature spec document (within the main doc), or in a related spec document e.g. 

```
0099-flux-capacitor.md  (requirements, description, background, acceptance criteria)
0099-flux-capacitor-api.md. (requirements, expected data, core AND non core api sections)
``` 

Having a 'Core API' ticket/specification does not make much sense other than a broad outline (in my opinion) because these APIs may and should be defined and grow with a feature of the Vega protocol.

**Comment by ashleyvega** [comment]
In GitLab by @barnabee on Oct 3, 2019, 14:47

@cdm that sounds neat but I’m not sure if it’s going to be quite as simple and clear cut as that:

- Some core APIs would be related to multiple features / specs at the protocol level, for example oracle submission, governance APIs, and even eventually submit market instruction (currently just orders) all are APIs that cross protocol features and/or core specs
- Non-core APIs may not relate to a protocol feature at all, though I guess that you just have an API spec on its own?
- Core APIs *might* work in different ways or even become less functional if we implement an optimisation in the core that means certain data is no longer available to the core APIs.

I’m not sure how you would handle these situations in your proposed model but it would be good to clarify howe you see it working that if that’s the route for the API specs?

**Comment by ashleyvega** [comment]
In GitLab by @cdm on Oct 3, 2019, 18:24

mentioned in merge request !15

**Comment by ashleyvega** [comment]
In GitLab by @cdm on Oct 3, 2019, 18:38

Thanks good questions @barnabee. All, I've now pushed a simplified version of the Core API feature spec md document (in the related MR branch !15).  

On the term 'core' I feel having spoken to several members of the team (including those who've worked on this product feature spec WIP), there's a bit of clarification on what is meant as 'core'. 

My view on what is core is listed in the definition in the md file, and @barnabee also mentioned a similar definition in a recent comment (https://gitlab.com/vega-protocol/product/issues/141#note_225620575).

The key thing for me is that it's not what API consumers think is CORE to their creation/operation of a bot/script/UI etc, as some of that is most probably generated by a read node and should be IMO. It is what is to provide a set of APIs that give access to the data available in the core using its current state, and typically not historic/auditing data or data that needs additional calculation.

I've gone through the list of domains/functions within the protocol and listed some simple functional requirements for this set of core APIs, however, as I suggested above it might make sense as the system grows to not have a single spec for all of them (we CAN keep this list of req's up to date as an alternative which could work too). The advantage of trying to separate APIs out or make them part of the core feature specs is that they are tied closely to the information or topic, rather than orphaned into a BIG-OL-LIST style spec. Especially if some of the APIs are classed core for a feature and some are classed as non-core e.g. history, audit, additional processing not already calculated by trading execution.

To take a stab at @barnabee's questions/feedback:

* Some core APIs would be related to multiple features / specs at the protocol level, for example oracle submission, governance APIs, and even eventually submit market instruction (currently just orders) all are APIs that cross protocol features and/or core specs

*This is a good point, so maybe it's worth keeping the BIG-OL-LIST style approach, but the negative is that it will grow large over time. I'd say a separate md file could be clearer where a Core API spans multiple features because it must be a special case, it must have a requirement to do something - need an example I guess*

* Non-core APIs may not relate to a protocol feature at all, though I guess that you just have an API spec on its own?

*Yes a non-core API would still need a feature spec document, so will have an md file*

* Core APIs *might* work in different ways or even become less functional if we implement an optimisation in the core that means certain data is no longer available to the core APIs.

*Not sure what this means, need clarification, if a Core API is deprecated need to change the Core API definition via the product ticket that triggered it* :dancer:
